packages/authorization-process/test/api/addClientPurpose.test.ts

```diff
@@ -87,107 +93,87 @@ describe("API /clients/{clientId}/purposes authorization test", () => {
-      .fn()
-      .mockRejectedValue(clientKindNotAllowed(mockClient.id));
-    const token = generateToken(authRole.ADMIN_ROLE);
-    const res = await makeRequest(token, mockClient.id);
-    expect(res.status).toBe(403);
-  });
-
-  it("Should return 403 for organizationNotAllowedOnClient", async () => {
-    authorizationService.addClientPurpose = vi
-      .fn()
-      .mockRejectedValue(
-        organizationNotAllowedOnClient(generateId(), mockClient.id)
-      );
-    const token = generateToken(authRole.ADMIN_ROLE);
-    const res = await makeRequest(token, mockClient.id);
-    expect(res.status).toBe(403);
-  });
+  it.each([
+    {
+      name: "clientNotFound",
+      error: clientNotFound(mockClient.id),
+      expectedStatus: 404,
+    },
+    {
+      name: "purposeNotFound",
+      error: purposeNotFound(mockPurpose.id),
+      expectedStatus: 404,
+    },
+    {
+      name: "noAgreementFoundInRequiredState",
+      error: noAgreementFoundInRequiredState(mockEservice.id, mockConsumerId),
+      expectedStatus: 400,
+    },
+    {
+      name: "noPurposeVersionsFoundInRequiredState",
+      error: noPurposeVersionsFoundInRequiredState(mockPurpose.id),
+      expectedStatus: 400,
+    },
+    {
+      name: "eserviceNotDelegableForClientAccess",
+      error: eserviceNotDelegableForClientAccess(mockEservice),
+      expectedStatus: 400,
+    },
+    {
+      name: "purposeAlreadyLinkedToClient",
+      error: purposeAlreadyLinkedToClient(mockPurpose.id, mockClient.id),
+      expectedStatus: 409,
+    },
+    {
+      name: "clientKindNotAllowed",
+      error: clientKindNotAllowed(mockClient.id),
+      expectedStatus: 403,
+    },
+    {
+      name: "organizationNotAllowedOnClient",
+      error: organizationNotAllowedOnClient(generateId(), mockClient.id),
+      expectedStatus: 403,
+    },
+    {
+      name: "organizationNotAllowedOnPurpose",
+      error: organizationNotAllowedOnPurpose(generateId(), mockPurpose.id),
+      expectedStatus: 403,
+    },
+    {
+      name: "purposeDelegationNotFound",
+      error: purposeDelegationNotFound(generateId()),
+      expectedStatus: 500,
+    },
+  ])(
+    "Should return $expectedStatus for $name",
+    async ({ error, expectedStatus }) => {
+      if (error) {
+        authorizationService.addClientPurpose = vi
+          .fn()
+          .mockRejectedValue(error);
+      }
+
+      const token = generateToken(authRole.ADMIN_ROLE);
+      const res = await makeRequest(token, mockClient.id);
+      expect(res.status).toBe(expectedStatus);
+    }
+  );
 
-  it("Should return 403 for organizationNotAllowedOnPurpose", async () => {
-    authorizationService.addClientPurpose = vi
-      .fn()
-      .mockRejectedValue(
-        organizationNotAllowedOnPurpose(generateId(), mockPurpose.id)
+  it.each([
+    {},
+    { clientId: "invalidId", purposeId: mockPurpose.id },
+    { clientId: mockClient.id, purposeId: "invalidId" },
+  ])(
+    "Should return 400 if passed invalid params",
```

Comment:
```suggestion
    "Should return 400 if passed invalid params: %s",
```

adding just `%s` seems to work to print the passed object in each case, just to understand more easily which test fails